### PR TITLE
Add whenP and unlessP

### PR DESCRIPTION
**Pull request by flintinatux**
I don't know about you, but sometimes I need a cleaner way to say that.

## to review:
- [ ] Read the codes.
- [ ] Wait for the green.
- [ ] Find an apropos squirrel.

### CODE REVIEW

**Comment by JWA111** [comment]
💯 bonus points for the custom meme

(now I'll review the code)

File: src/unlessP.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+const curry  = require('ramda/src/curry')
+const ifElse = require('ramda/src/ifElse')
+
+// unlessP :: (a -> Promise Boolean) -> (a -> Promise a) -> a -> Promise a
+const unlessP = (pred, f, x) => {
+  const run = () =>
+    Promise.resolve(x).then(f)
+
+  return Promise.resolve(x).then(pred)
+    .then(ifElse(Boolean, always(x), run))
```
Comment (mgreystone) [inline src/unlessP.js]: I guess you could just do `whenP(complement(pred), f, x)`, if you felt so inclined...

Comment (flintinatux) [reply src/unlessP.js]: I thought about that, but the predicate returns a `Promise`, not a `Boolean`.

File: API.md [hunk 2]
```diff
@@ -618,6 +620,25 @@ Runs the given function with the supplied value, and then resolves with that val
...
+
+`@articulate/funky/lib/unlessP`
+
+```haskell
+unlessP :: (a -> Promise Boolean) -> (a -> Promise a) -> a -> Promise a
+```
+
+An async version of [`R.unless`](http://devdocs.io/ramda/index#unless) that accepts Promise-returning functions.
+
+Tests a value with an async predicate.  If the predicate resolves truthy, it resolves with the original value.  If the predicate resolves falsey, it resolves with the result of calling the supplied function.
```
Comment (rpearce) [inline API.md]:
```suggestion
Tests a value with an async predicate.  If the predicate resolves truthy, it resolves with the original value.  If the predicate resolves falsy, it resolves with the result of calling the supplied function.
```

File: API.md [hunk 3]
```diff
@@ -680,3 +701,22 @@ const schema = Joi.object({
...
+
+`@articulate/funky/lib/whenP`
+
+```haskell
+whenP :: (a -> Promise Boolean) -> (a -> Promise a) -> a -> Promise a
+```
+
+An async version of [`R.when`](http://devdocs.io/ramda/index#when) that accepts Promise-returning functions.
+
+Tests a value with an async predicate.  If the predicate resolves truthy, it resolves with the result of calling the supplied function.  If the predicate resolves falsey, it resolves with the original value.
```
Comment (rpearce) [inline API.md]:
```suggestion
Tests a value with an async predicate.  If the predicate resolves truthy, it resolves with the result of calling the supplied function.  If the predicate resolves falsy, it resolves with the original value.
```

File: test/unlessP.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    it('does not call the supplied function', () =>
+      expect(addSpy.calls.length).to.equal(0)
+    )
+
+    it('resolves with the original value', () =>
+      expect(res()).to.equal(2)
+    )
+  })
+
+  describe('when the predicate resolves with a falsey value', () => {
```
Comment (rpearce) [inline test/unlessP.js]:
```suggestion
  describe('when the predicate resolves with a falsy value', () => {
```

File: test/whenP.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    it('calls the supplied function', () =>
+      expect(addSpy.calls[0]).to.eql([ 1, 2 ])
+    )
+
+    it('resolves with the result of the supplied function', () =>
+      expect(res()).to.equal(3)
+    )
+  })
+
+  describe('when the predicate resolves with a falsey value', () => {
```
Comment (rpearce) [inline test/whenP.js]:
```suggestion
  describe('when the predicate resolves with a falsy value', () => {
```

**Comment by flintinatux** [comment]
The angry red ❌ is because Coveralls is down.  I see no need to wait on that.

squirrel

File: src/unlessP.js (hunk 1 above)
Comment (pklingem) [reply src/unlessP.js]: `complementP`! :troll:

**Pull request merged**